### PR TITLE
Doc and Error Message Improvements

### LINKED ISSUE #7659: Convert the box decoration border radius with non-uniform border assert into a detailed FlutterError
## Steps to reproduce

Run the provided code. You'll see widgets on a pink background.
Uncomment the `borderRadius:` line and reload.
The pink box now has rounded borders, but that's all you see. The widgets are now obscured.

## Example code

```dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart' show debugPaintSizeEnabled;

void main() {
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    //debugPaintSizeEnabled = true;
    return new Scaffold(
        appBar: new AppBar(
          title: new Text(config.title),
        ),
        body: new Center(
            child: new Container(
                padding: new EdgeInsets.all(8.0),
                decoration: new BoxDecoration(
                 //borderRadius: new BorderRadius.all(const Radius.circular(8.0)),
                  backgroundColor: Colors.pink[500],
                  border: new Border(
                      top: new BorderSide(
                          color: new Color.fromARGB(255, 30, 255, 30),
                          width: 4.0)),
                ),
                child: new Row(children: <Widget>[
                  new Container(
                    padding: new EdgeInsets.all(4.0),
                      constraints: new BoxConstraints(maxWidth: 35.0),
                      child: new Image.network(
                          'http://4.bp.blogspot.com/-Q4DRkyhsKwI/U9SsY-ktENI/AAAAAAABIQw/rgMWScAzIQo/s1600/Fuchsia+flowers+1.JPG')),
                  new Container(
                    child: new RichText(
                      text: new TextSpan(
                        text: 'Fuschia ',
                        style: new TextStyle(
                          color: Colors
                              .white, //new Color.fromARGB(255, 255, 255, 255),
                          decoration: TextDecoration.underline,
                          decorationColor: Colors
                              .blue[500], //new Color.fromARGB(255, 0, 0, 255),
                          fontStyle: FontStyle.italic,
                        ),
                        children: <TextSpan>[
                          new TextSpan(
                              text: 'IS',
                              style:
                                  new TextStyle(fontWeight: FontWeight.w900)),
                          new TextSpan(text: ' the new black!'),
                        ],
                      ),
                    ),
                  )
                ]))));
  }
}
```

## Flutter Doctor

```
$ flutter doctor
[✓] Flutter (on Mac OS, channel master)
    • Flutter at /Users/shaza/GoogleDrive/TMP-flutter/flutter
    • Framework revision 3cde53590c (13 hours ago), 2017-01-25 19:52:18
    • Engine revision fb69728f33
    • Tools Dart version 1.22.0-dev.9.1

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.0)
    • Android SDK at /Users/shaza/Library/Android/sdk
    • Platform android-25, build-tools 25.0.0
    • Java(TM) SE Runtime Environment (build 1.8.0_112-b16)

[✓] iOS toolchain - develop for iOS devices (Xcode 7.3.1)
    • XCode at /Applications/Xcode.app/Contents/Developer
    • Xcode 7.3.1, Build version 7D1014

[✓] IntelliJ IDEA Community Edition (version 2016.3.1)
    • Dart plugin version 163.9166.22
    • Flutter plugin version 0.1.8

[✓] Connected devices
    • iPhone 6s • DFD0512F-17D2-4828-8CEA-A98DBDB5499E • ios • iOS 9.3 (simulator)
```

### LINKED ISSUE #7703: Poor messaging for assert when getting local coordinates of RenderBox: 'package:flutter/src/rendering/box.dart': Failed assertion: line 1342
```
16:28:26.258: I/flutter (13143): ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
16:28:26.262: I/flutter (13143): The following assertion was thrown building ArmadilloDragTarget<StoryClusterId>(dirty; state:
16:28:26.265: I/flutter (13143): _DragTargetState<StoryClusterId>(564928223)):
16:28:26.267: I/flutter (13143): 'package:flutter/src/rendering/box.dart': Failed assertion: line 1342: 'debugDoingThisResize ||
16:28:26.270: I/flutter (13143): debugDoingThisLayout ||
16:28:26.273: I/flutter (13143):                   (RenderObject.debugActiveLayout == parent && _size._canBeUsedByParent)': is not
16:28:26.276: I/flutter (13143):                   true.
16:28:26.278: I/flutter (13143):
16:28:26.281: I/flutter (13143): Either the assertion indicates an error in the framework itself, or we should provide substantially
16:28:26.284: I/flutter (13143): more information in this error message to help you determine and fix the underlying cause.
16:28:26.287: I/flutter (13143): In either case, please report this assertion by filing a bug on GitHub:
16:28:26.289: I/flutter (13143):   https://github.com/flutter/flutter/issues/new
16:28:26.292: I/flutter (13143):
16:28:26.294: I/flutter (13143): When the exception was thrown, this was the stack:
16:28:26.297: I/flutter (13143): #2      RenderBox.size.<anonymous closure> (package:flutter/src/rendering/box.dart:1342)
16:28:26.300: I/flutter (13143): #4      RenderBox.size (package:flutter/src/rendering/box.dart:1332)
16:28:26.303: I/flutter (13143): #5      RenderTransform._effectiveTransform (package:flutter/src/rendering/proxy_box.dart:1428)
16:28:26.305: I/flutter (13143): #6      RenderTransform.applyPaintTransform (package:flutter/src/rendering/proxy_box.dart:1469)
16:28:26.308: I/flutter (13143): #7      RenderBox._collectPaintTransform (package:flutter/src/rendering/box.dart:1769)
16:28:26.311: I/flutter (13143): #8      RenderBox.globalToLocal (package:flutter/src/rendering/box.dart:1779)
16:28:26.314: I/flutter (13143): #9      PanelDragTargetsState.build.<anonymous closure>.<anonymous closure> (/usr/local/google/home/anwilson/fuchsia/apps/sysui/armadillo/lib/panel_drag_targets.dart:297)
16:28:26.316: I/flutter (13143): #10     Iterable.forEach (dart:core/iterable.dart:226)
16:28:26.319: I/flutter (13143): #11     PanelDragTargetsState.build.<anonymous closure> (/usr/local/google/home/anwilson/fuchsia/apps/sysui/armadillo/lib/panel_drag_targets.dart:293)
16:28:26.322: I/flutter (13143): #12     _DragTargetState.build (/usr/local/google/home/anwilson/fuchsia/apps/sysui/armadillo/lib/armadillo_drag_target.dart:295)
16:28:26.324: I/flutter (13143): #13     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2872)
16:28:26.327: I/flutter (13143): #14     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2767)
16:28:26.329: I/flutter (13143): #15     StatefulElement.update (package:flutter/src/widgets/framework.dart:3009)
16:28:26.332: I/flutter (13143): #16     Element.updateChild (package:flutter/src/widgets/framework.dart:2050)
16:28:26.335: I/flutter (13143): #17     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:2884)
16:28:26.338: I/flutter (13143): #18     BuildableElement.rebuild (package:flutter/src/widgets/framework.dart:2767)
16:28:26.340: I/flutter (13143): #19     StatefulElement.update (package:flutter/src/widgets/framework.dart:3009)
16:28:26.343: I/flutter (13143): #20     Element.updateChild (package:flutter/src/widgets/framework.dart:2050)
16:28:26.346: I/flutter (13143): #21     _LayoutBuilderElement._layout.<anonymous closure> (package:flutter/src/widgets/layout_builder.dart:120)
16:28:26.348: I/flutter (13143): #22     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:1731)
16:28:26.351: I/flutter (13143): #23     _LayoutBuilderElement._layout (package:flutter/src/widgets/layout_builder.dart:108)
16:28:26.354: I/flutter (13143): #24     RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:1833)
16:28:26.357: I/flutter (13143): #25     PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:975)
16:28:26.360: I/flutter (13143): #26     RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:1833)
16:28:26.362: I/flutter (13143): #27     _RenderLayoutBuilder.performLayout (package:flutter/src/widgets/layout_builder.dart:207)
16:28:26.365: I/flutter (13143): #28     RenderObject.layout (package:flutter/src/rendering/object.dart:1735)
16:28:26.368: I/flutter (13143): #29     RenderFlex.performLayout (package:flutter/src/rendering/flex.dart:536)
16:28:26.374: I/flutter (13143): #30     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1611)
16:28:26.377: I/flutter (13143): #31     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:953)
16:28:26.380: I/flutter (13143): #32     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.beginFrame (package:flutter/src/rendering/binding.dart:220)
16:28:26.382: I/flutter (13143): #33     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.beginFrame (package:flutter/src/widgets/binding.dart:314)
16:28:26.385: I/flutter (13143): #34     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:165)
16:28:26.388: I/flutter (13143): #35     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:652)
16:28:26.390: I/flutter (13143): #36     BindingBase&SchedulerBinding.handleBeginFrame (package:flutter/src/scheduler/binding.dart:587)
16:28:26.393: I/flutter (13143): #37     _beginFrame (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:83)
16:28:26.396: I/flutter (13143): (elided 3 frames from class _AssertionError)
16:28:26.398: I/flutter (13143): ════════════════════════════════════════════════════════════════════════════════════════════════════
```

### LINKED ISSUE #13013: Improve documentation and assert for Stack.textDirection; also, API docs for Localizations should admit to adding a Directionality
## Steps to Reproduce
The following empty Stack layout trips an assertion.
```dart
import 'package:flutter/widgets.dart';

void main() {
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  MyApp() {
  }

  @override
  Widget build(BuildContext context) {
    return new Stack();
  }
}
```
## Logs
```
I/flutter (24950): ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
I/flutter (24950): The following assertion was thrown during performLayout():
I/flutter (24950): 'package:flutter/src/painting/alignment.dart': Failed assertion: line 513 pos 12: 'direction !=
I/flutter (24950): null': is not true.
I/flutter (24950): 
I/flutter (24950): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter (24950): more information in this error message to help you determine and fix the underlying cause.
I/flutter (24950): In either case, please report this assertion by filing a bug on GitHub:
I/flutter (24950):   https://github.com/flutter/flutter/issues/new
I/flutter (24950): 
I/flutter (24950): When the exception was thrown, this was the stack:
I/flutter (24950): #2      AlignmentDirectional.resolve (package:flutter/src/painting/alignment.dart:513:12)
I/flutter (24950): #3      RenderStack._resolve (package:flutter/src/rendering/stack.dart:327:36)
I/flutter (24950): #4      RenderStack.performLayout (package:flutter/src/rendering/stack.dart:436:5)
I/flutter (24950): #5      RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
I/flutter (24950): #6      RenderView.performLayout (package:flutter/src/rendering/view.dart:142:13)
I/flutter (24950): #7      RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1837:7)
I/flutter (24950): #8      PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1126:18)
I/flutter (24950): #9      BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:262:19)
I/flutter (24950): #10     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:581:22)
I/flutter (24950): #11     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:200:5)
I/flutter (24950): #12     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:713:15)
I/flutter (24950): #13     BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:649:9)
I/flutter (24950): #14     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/rendering/binding.dart:286:20)
I/flutter (24950): #16     _Timer._runTimers (dart:isolate-patch/dart:isolate/timer_impl.dart:367)
I/flutter (24950): #17     _Timer._handleMessage (dart:isolate-patch/dart:isolate/timer_impl.dart:401)
I/flutter (24950): #18     _RawReceivePortImpl._handleMessage (dart:isolate-patch/dart:isolate/isolate_patch.dart:163)
I/flutter (24950): (elided 3 frames from class _AssertionError and package dart:async-patch)
I/flutter (24950): 
I/flutter (24950): The following RenderObject was being processed when the exception was fired:
I/flutter (24950):   RenderStack#46ac0 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (24950):   creator: Stack ← MyApp ← [root]
I/flutter (24950):   parentData: <none>
I/flutter (24950):   constraints: BoxConstraints(w=1024.0, h=768.0)
I/flutter (24950):   size: MISSING
I/flutter (24950):   alignment: AlignmentDirectional.topStart
I/flutter (24950):   textDirection: null
I/flutter (24950):   fit: loose
I/flutter (24950):   overflow: clip
I/flutter (24950): This RenderObject has no descendants.
I/flutter (24950): ════════════════════════════════════════════════════════════════════════════════════════════════════
```

## Flutter Doctor
```
[✓] Flutter (on Mac OS X 10.13.1 17B48, locale en-CA, channel alpha)
    • Flutter at /Users//flutter/flutter
    • Framework revision e8aa40eddd (4 weeks ago), 2017-10-17 15:42:40 -0700
    • Engine revision 7c4142808c
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users//Library/Android/sdk
    • Platform android-25, build-tools 25.0.3
    • ANDROID_HOME = /Users//Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.1, Build version 9B55
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.2.5)
    • Flutter plugin version 18.2
    • Dart plugin version 172.4155.35

[✓] Connected devices
    • SM T810 • 33003b44d0cd3317 • android-arm • Android 7.0 (API 24) (emulator)
```

### LINKED ISSUE #13861: Better error message for sliver grid
I'm getting a NaN in sliver_grid.dart.

```
I/flutter (13931): ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
I/flutter (13931): The following UnsupportedError was thrown during performLayout():
I/flutter (13931): Unsupported operation: Infinity or NaN toInt
I/flutter (13931): 
I/flutter (13931): When the exception was thrown, this was the stack:
I/flutter (13931): #0      _Double.toInt (dart:core-patch/dart:core/double.dart:151)
I/flutter (13931): #1      _Double.ceil (dart:core-patch/dart:core/double.dart:126)
I/flutter (13931): #2      SliverGridDelegateWithMaxCrossAxisExtent.getLayout (package:flutter/src/rendering/sliver_grid.dart:425)
I/flutter (13931): #3      RenderSliverGrid.performLayout (package:flutter/src/rendering/sliver_grid.dart:522)

```

### LINKED ISSUE #14842: Unexpected top padding in ListView put inside scaffold with no appBar
If you put `ListView` somewhere inside `Scaffold` having no `appBar`, `ListView` has top padding. There is no padding if `appBar` is present. There is no padding if you specify `EdgeInsets.zero` padding for `ListView` explicitly. See following code snipped for example   

```dart
...
    return new MaterialApp(
      home: new Scaffold(
//        appBar: new AppBar(
//          title: new Text("test"),
//        ),
        body: new Column(
          children: <Widget>[
            new Container(
              child: new Center(
                  child: new Text("some top widget")
              ),
              decoration: new BoxDecoration(
                color: Colors.red,
              ),
              height: 100.0,
            ),
            new Expanded(
              child: new ListView(
                children: items,
//                padding: EdgeInsets.zero, // explicit padding as a workaround
              ),
            ),
          ],
        ),
      ),
    );
...
```
See screenshot:

bug

## Flutter Doctor:
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (on Linux, locale en_US.UTF-8, channel master)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] Android Studio (version 3.0)
[!] IntelliJ IDEA Ultimate Edition (version 2017.3)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.20.1)
[✓] Connected devices
```

### LINKED ISSUE #15423: Improve the ListTile docs: explain the Text maxLines dependency
Flutter beta.

Docs say `ListTiles` are all the same height, and all 1–3 lines high. When I build `ListTiles` with subtitle property containing a very length Text object, those `ListTiles` are *wayyyy* taller than three lines and certainly not of equal height like the docs say they're supposed to (even if you construe "three lines" as "three spans of text separated by \n".

See screenshot. These are two ListTiles.

screen shot 2018-03-11 at 23 07 49

Relevant code:
```dart
    new ListView(
        shrinkWrap: true,
        scrollDirection: Axis.vertical,
        ListTile.divideTiles(
            color: Colors.deepOrange,
            tiles: items.map((item) {
                return _buildRow(item);
            }).toList()
        ).toList()
    )

    Widget _buildRow(Item item) {
        return new Dismissible(
            key: new Key(item.id.toString()),
            //...
            child: new Padding(
                child: new ListTile(
                    title: new Text(item.title),
                    subtitle: new Text(item.content),
                    textAlign: TextAlign.justify
                )
                //...
            )
        );
    }
```

### LINKED ISSUE #19445: Document that widgets in the overflow of stack do not respond to gestures
Internal: b/143100923

When a `Stack` contains `MyWidget` inside of a `Positioned`.
```dart
    Stack(
      overflow: Overflow.visible,
      children: [
        Positioned(
        top: -50.0,
        child: MyWidget(),
      )],
    );
```

Since overflow is `Overflow.visible`, `MyWidget` displays outside of the `Stack`. However, it doesn't accept gestures in the overflowed area. I have no idea if it's easy, difficult or downright impossible to make this change, but I can think of a gazillion use cases for that, which at the moment must be satisfied by using the much more complex Overlay, or some Stack tricks.

And in case it's impossible to make this change, then at least the documentation should clearly state that overflowed content can't receive gestures.

Also, see: https://stackoverflow.com/questions/51366761/in-flutter-how-can-a-positioned-widget-feel-taps-outside-of-its-parent-stack-ar

### LINKED ISSUE #27386: AnimatedSize needs an example
Especially because of the `vsync` argument.

### LINKED ISSUE #32701: CupertinoButton in CupertinoNavigationBar as trailing gets cut
Adding a CupertinoButton with Text to CupertinoNavigationBar as trailing cuts of the Text on its half height.

## Steps to Reproduce

https://gist.github.com/tb2761/567623b671e5d15cfadfb1a27248fcf4

## Logs

```
[√] Flutter (Channel stable, v1.5.4-hotfix.2, on Microsoft Windows [Version 10.0.17763.475], locale de-DE)
    • Flutter version 1.5.4-hotfix.2 at C:\Users\Tim\Flutter\src\flutter
    • Framework revision 7a4c33425d (2 weeks ago), 2019-04-29 11:05:24 -0700
    • Engine revision 52c7a1e849
    • Dart version 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\Tim\AppData\Local\Android\Sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = C:\Users\Tim\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.4)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 35.3.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[√] VS Code (version 1.33.1)
    • VS Code at C:\Users\Tim\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.0.2

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 9 (API 28) (emulator)

• No issues found!
```

### LINKED ISSUE #44128: Assertion failed: "onGenerateRoute" shows assertion when I try to go the same route for the 2nd time.
Condition:
- using Chrome-Develop for the web

- using "flutter channel master"  2019-11-5 tokyo 
("flutter version" shows tons of v 1.XX.XX . so I am not sure what to write here)

Step:

1. With the code below, start the app by "flutter run -d chrome"
2.Click "to the Home" button.
3.Click "to the Root" button.
4.You see the Red Screen with the assertion message.

Comment:

- as you can see from the code, I have tested using onGenerateRoute with the switch case. This works.

- but what I want to achieve is to create routing table from setting file. which probably need Map

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  final Map rootMap;
  MyApp():
    rootMap = {
      "/": MaterialPageRoute(builder:(context) => Root()),
      '/Home': MaterialPageRoute(builder:(context) => Home()),
    };

  Route<dynamic> generateRouteByMap(RouteSettings settings) {
    if(rootMap[settings.name] != null){
      return rootMap[settings.name];
    }
    else{
      return MaterialPageRoute(builder:(context) => Root());
    }
  }

  Route<dynamic> generateRouteBySwitch(RouteSettings settings) {
    switch(settings.name){
      case "/Home":
        return MaterialPageRoute(builder:(context) => Home());
      default:
        return MaterialPageRoute(builder:(context) => Root());
    }
  }
  
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Reproduce',
        // routes: {
        //   "/": (context) => Root(),
        //   "/Home": (context) => Home(),
        // },
      //  onGenerateRoute: generateRouteBySwitch,
      onGenerateRoute: generateRouteByMap,
    );
  }
}

class Root extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Row(
      children: <Widget>[
        Text("Root"),
        FlatButton(
          child: Text("to the Home"),
          onPressed: () {
            Navigator.popAndPushNamed(context, "/Home");
          },
        ),
      ],
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Row(
      children: <Widget>[
        Text("home"),
        FlatButton(
          child: Text("to The Root"),
          onPressed: () {
            Navigator.popAndPushNamed(context, "/");
          },
        ),
      ],
    );
  }
}
````
and the Assertion message is following
══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
The following assertion was thrown while handling a gesture:
Assertion failed: org-dartlang-app:///packages/flutter/src/widgets/navigator.dart:1788:12
route._navigator == null
is not true
package:flutter/src/gestures/binding.dart 156:7
[_handlePointerEvent]
package:flutter/src/gestures/binding.dart 102:7
[_flushPointerEventQueue]
package:flutter/src/gestures/binding.dart 86:32
[_handlePointerDataPacket]
package:build_web_compilers/src/dev_compiler/C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/pointer_binding.dart 89:15
[_onPointerData]
package:build_web_compilers/src/dev_compiler/C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/pointer_binding.dart 261:7                 <fn>
package:build_web_compilers/src/dev_compiler/C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/pointer_binding.dart 173:16                <fn>

Handler: "onTap"
Recognizer:
TapGestureRecognizer#0a0fa
════════════════════════════════════════════════════════════════════════════════════════════════════
Another exception was thrown: Assertion failed: org-dartlang-app:///packages/flutter/src/widgets/navigator.dart:2330:12

### PR DESCRIPTION
**Pull request by Piinks**
## Description

A couple of weeks ago I did a quarantine doc sprint. 😷  These are the remnants that weren't merged in other PRs.
The goal was to de-clutter the issue tracker of highly upvoted documentation/error message bugs, or those that are also labeled as regressions and crashers. 

This may be broken out further into more PRs for clarity. :)

## Related Issues

Fixes #19445
Fixes #7659
Fixes #44128
Fixes #13861
Fixes #7703
Fixes #13013
Fixes #14842
Fixes https://github.com/flutter/flutter/issues/15423
Fixes https://github.com/flutter/flutter/issues/58371
Fixes https://github.com/flutter/flutter/issues/47764
Fixes https://github.com/flutter/flutter/issues/32701
Fixes https://github.com/flutter/flutter/issues/27386

## Tests

Just doc changes or assertion refactoring for better messaging.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/box_border.dart [hunk 1]
```diff
@@ -394,24 +394,42 @@ class Border extends BoxBorder {
...
 
+  bool get _widthIsUniform {
     final double topWidth = top.width;
     if (right.width != topWidth ||
-        bottom.width != topWidth ||
-        left.width != topWidth)
+      bottom.width != topWidth ||
+      left.width != topWidth)
       return false;
+    return true;
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/box_border.dart:424]: Can this (and the other two above and below) be written in a single return statement:

```suggestion
    return right.width == topWidth && bottom.width == topWidth && left.width == topWidth;
```

(Please double check that I transformed the logic correctly...

File: packages/flutter/lib/src/painting/box_border.dart [hunk 2]
```diff
@@ -522,8 +540,30 @@ class Border extends BoxBorder {
...
-    assert(borderRadius == null, 'A borderRadius can only be given for uniform borders.');
-    assert(shape == BoxShape.rectangle, 'A border can only be drawn as a circle if it is uniform.');
+    assert(() {
+      if (borderRadius != null) {
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary('A borderRadius can only be given for a uniform Border.'),
+          if (!_colorIsUniform) ErrorDescription('BorderSide.color'),
+          if (!_widthIsUniform) ErrorDescription('BorderSide.width'),
+          if (!_styleIsUniform) ErrorDescription('BorderSide.style'),
+          ErrorDescription('is not uniform.'),
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/box_border.dart:550]: This error will be grammatically incorrect (is vs. are) if more than one property is not uniform (assuming that can happen).

File: packages/flutter/lib/src/painting/box_border.dart [hunk 3]
```diff
@@ -522,8 +540,30 @@ class Border extends BoxBorder {
...
+      return true;
+    }());
+    assert(() {
+      if (shape != BoxShape.rectangle) {
+        throw FlutterError.fromParts(<DiagnosticsNode>[
+          ErrorSummary('A Border can only be drawn as a circle if it is uniform'),
+          if (!_colorIsUniform) ErrorDescription('BorderSide.color'),
+          if (!_widthIsUniform) ErrorDescription('BorderSide.width'),
+          if (!_styleIsUniform) ErrorDescription('BorderSide.style'),
+          ErrorDescription('is not uniform.'),
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/box_border.dart:562]: same here.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -3285,6 +3285,9 @@ class Stack extends MultiChildRenderObjectWidget {
   /// Some children in a stack might overflow its box. When this flag is set to
   /// [Overflow.clip], children cannot paint outside of the stack's box.
   ///
+  /// When set to [Overflow.visible], the visible overflow area wil not accept
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3288]: Can you also add this information to the doc of Overflow.visible?

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 5]
```diff
@@ -2404,7 +2404,11 @@ class _RouteEntry extends RouteTransitionRecord {
     assert(currentState == _RouteLifecycle.push || currentState == _RouteLifecycle.pushReplace || currentState == _RouteLifecycle.replace);
     assert(navigator != null);
     assert(navigator._debugLocked);
-    assert(route._navigator == null);
+    assert(
+      route._navigator == null,
+      'The pushed route has already been used. When pushing a route, a new '
+      'Route object should be provided.',
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/navigator.dart:2410]: should be -> must be

**Review by goderbauer — COMMENTED** [review]
This is great. Thanks for the clean-up!

**Review by allanbatista-vwi — APPROVED** [review]

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**